README
vdr-ttxtsubs - a teletext subtitle plugin for the Linux Video Disk Recorder
===========================================================================

About
-----

Description:
  This plug-in implements displaying, recording and replaying teletext
  based subtitles using the on screen display.

Current Maintainer:
  Tobias Grimm <vdr@e-tobi.net>

Original Author:
  Ragnar Sundblad <ragge@nada.kth.se>

Homepage:
  http://projects.vdr-developer.org/projects/show/plg-ttxtsubs

Old homepage:
  http://www.nada.kth.se/~ragge/vdr/ttxtsubs/

Copyright:
  (C) 2003 - 2008 Ragnar Sundblad
  (C) 2009 - 2010 Tobias Grimm

License:
  This program is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the
  Free Software Foundation; either version 2 of the License, or (at your
  option) any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
  more details.

  You should have received a copy of the GNU General Public License along with
  this program (see file `COPYING`); if not, write to the
  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
  MA 02110-1301 USA


Introduction
------------

This plug-in is for VDR >= 1.7.6. It currently needs a some changes to the VDR
code in order to work. A patch file is included in patches/ which can be safely
applied even if you don't use the ttxtsubs plug-in. For your convenience, the
patch is also available as a set of logical chunks, each covering a specific
feature.

The main responsibilities of the VDR patch are to enable the reception of the
teletext data stream and to discover the subtitle pages. So even without the
plug-in, the patch will allow you to record teletext subtitles, which e.g. can
be shown in 3'rd partly players like VLC.


Usage
-----

Important: In order to be able to receive the teletext subtitles, you must
enable the teletext subtitle support in VDR's setup. You will find this setting
in the DVB section of the setup.

Ideally the available subtitle pages / languages will be auto-detected. But this
might not be possible with all broadcasters. Especially the German broadcasters
don't care very much about the DVB standard in this regard and you have to
configure the subtitle page(s) for each channel manually.

In order to do so, you have to stop VDR and edit the channels.conf. After the
Tpid entry, you can add tuples of language codes and page numbers like this:

    ...:2001;150=deu,151=fin:...

See `man 5 vdr` for a detailed description of the channels.conf.

Future versions of the plug-in will provide an OSD setup for this setting, but
right now you are on your own.

To see the subtitles, you also must select the subtitle languages you would
like to see in the plug-ins OSD setup. In the future this might get somehow
merged with VDRs built-in DVB-subtitle selection.

In the plug-ins setup you can also configure some parameters which control how
the subtitles will be displayed.


Additional notes
----------------

A Note about reported problems with Digitainer II:

If you own a Digitainer II, you may run into an issue with the OSD and a
corrupted main menu. OlvierH reported, that the following setting at leas
partially solves this:

Advanced Chipset Features - AGP &P2P Bridge Control - VGA Share Memory Size: 64M
(default is 32M)


HISTORY
VDR Plugin 'ttxtsubs' Revision History
--------------------------------------

2013-03-10: Version 0.3.0
- Removed obsolete (unused since 0.2.0) setup options: DVB Source Selection,
  Live Delay, Record Subtitles, and Workaround for some French chns.
- Added hard-coded workaround for french channels where page numbes 850-859
  must be mapped to 880-889 (Closes #865, #703) (Thx to Jussi Tepponen)
- Some code refactorings
- Adapt to VDR 1.7.26 API changes.
- Updated patch for VDR 1.7.38
- Sync subtitles using PTS (Closes #119) (Thx to Matti Horila and Dimitar Petrovski)
- Added option for setting the left/center/right justification of subtitles
  (Closes #957) (Thx to Jussi Tepponen)

2011-10-08: Version 0.2.4
- Updated patch for VDR 1.7.21 (Closes #745)

  !!! With this release the channels.conf format has changed !!!
  !!! You may need to manually update your channels.conf     !!!

  With VDR 1.7.21 the subtitle PIDs are stored in the channels.conf as an
  extension to the teletext PID (see vdr.5 for details).

  This change made it necessary to change the extension of the tpid in
  channels.conf used by the ttxtsubs patch:

  old: ...:201;150=deu,151=fin:...
  new  ...:201+150=deu,151=fin;2001,2002:...

  ...where 201 is the TPID, 150 and 151 are the teletext pages for the
  German and Finnish teletext subtitles and 2001,2002 are the DVB
  subtitle PIDs.

2011-08-21: Version 0.2.3
- Made changes in font settings be applied instantly, without the need to
  change the channel (patch provided by Rolf Ahrenberg)
- Converted *.po to UTF-8
- Updated patch for VDR 1.7.20 (Closes #689)
- Don't log all page info received from VDR (Closes #331)
- Show subtitles with backround rectangle, if outline width is < 2
  (patch provided by Mike Lampard) (Closes #247)

2010-05-08: Version 0.2.2
- Fixed channel retune triggering in VDR patch and updated patch to 1.7.14
  (Dropped 1.7.12 and 1.7.13 patch)
- Got rid of the backround thread
- Fixed live TV subtitles on FF cards, where there is no transfer mode
- Updated README and TROUBLESHOOTING (Closes #323)
- Increased internal frame buffer and log buffer overflows (Closes #295)

2010-03-06: Version 0.2.1
- Added Estonian translation by Arthur Konovalov (Closes #271)
- Finnish translation updated by Rolf Ahrenberg
- Trigger OSD update when an empty page is sent (making the previous subtitle
  disappear) (Closes #275)
- French translation updated by Michaël Nival (Closes #278)
- Makefile adjustments for VDR 1.7.13.
- Fixed text rendering to not cut off bottom part (Closes #276)
- Italian translation update by Diego Pierotto (Closes #283)
- Added patch for VDR 1.7.13 (Closes #284)

2010-02-27: Version 0.2.0
- Add Ukrainian translation by Yarema aka Knedlyk (Closes #130)
- Some code refactoring
- Major VDR patch update for 1.7.12 (Closes #236, References #139)) - The patch
  now covers the following changes to VDR:
   - Record teletext subtitles
   - Added setup option to enable teletext subtitle recording
   - Allow configuration of fixed teletetxt subtitle pages
   - Capture teletext subtitle pages from PMT
   - Pass teletext date ttxtsubs plugin
  Special Thx to Rolf Ahrenberg!
- Fixed replay when the current live ttxtsubs page differs from the
  one in the recording - Thx to Rolf Ahrenberg (Closes #139)
- Show colored subtitles if color information is provided e.g. for
  different speakers and dropped custom color settings (Closes #61)
- Position subtitles always centered at the bottom, independent of the
  OSD resolution and dropped custom text position settings (Closes #231)
- Render subtitles outlined (Closes #51)
- Decrypt encrypted teletext stream (Closes #241)
- Don't redisplay unchanged subtitle pages

2009-04-26: Version 0.1.0
- Updated VDR patch and plugin to VDR 1.7.6 with support for the TS recording
  format and added separate delay setting for TS recordings. Older VDR versions
  are no longer supported. Credits go to Rolf Ahrenberg - Thanks a lot!
- Silence compiler warning about unchecked read return value.
  (thx to Rolf Ahrenberg)
- After replaying, reset the last channel variable and trigger a channel switch
  to reinitialize the live ttxtsubs display. (thx to Rolf Ahrenberg)
- Removed unused local copy of vdrttxtsubshooks.c|h

2009-04-03: Version 0.0.9
- Updated French translation provided by Michaël Nival (Closes #49)
- Install plug-in to lib dir with --remove-destination
- Added note in README about reported problems with Digitainer II (Closes #62)
- Don't show page selection menu, when there is no current channel (Closes #111)

2009-01-04: Version 0.0.8
- Updated Italien translation provided by Diego Pierotto
- Updated Russian translation provided by Oleg Roitburd (Closes #47)
- Fixed displaying of multiple rows when antialiasing is enabled by acquiring a
  single OSD area for all rows - Thx to Rolf Ahrenberg (Closes #24)

2008-12-18: Version 0.0.7
- Try to set OSDArea to 8 bpp when AntiAliasing is enabled (Closes #17)

2008-12-14: Version 0.0.6
- Project moved to http://projects.vdr-developer.org/projects/show/plg-ttxtsubs
- Applied patch from Rolf Ahrenberg (vdr-ttxtsubs-0.0.5-raastinrauta-edition.diff.gz)
  which introduces the following changes:
  - Rolf Ahrenberg
    - Updated the core patch and Makefile against VDR's current API
    - Added font, color, and DVB source selection
    - Added UTF8 support
    - Added layered OSD support
    - Fixed EPG scan hiccups
    - Fixed several race conditions in destructors
    - Silenced compilation warnings
  - Vyacheslav Diconov
    - Russian translation
  - Ville Skyttä
    - fixed crash in plugin shutdown
    - cleaned up logging to use syslog
  - Thomas Günther
    - German translation
  - Michaël Nival
    - Added missing const declarations
    - French translation
  - Luca Olivetti
    - Fixed displaying subtitles setting
    - Added a thread based subtitles scanner
    - Added delay settings
   - Tobias Grimm
    - Added manual page selection mode

siinfo.c
/*
 * vdr-ttxtsubs - A plugin for the Linux Video Disk Recorder
 * Copyright (c) 2003 - 2008 Ragnar Sundblad <ragge@nada.kth.se>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

#include <sys/ioctl.h>
#include <stdio.h>
#include <stdint.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <time.h>
#include <sys/poll.h>
#include <time.h>
#include <netinet/in.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <map>

#include <vdr/device.h>
#include <vdr/dvbdevice.h>
#include <vdr/channels.h>

#include "linux/dvb/dmx.h"
#include "siinfo.h"

#include "ttxtsubsglobals.h"
#include "ttxtsubschannelsettings.h"

#define DESCR_TELETEXT 0x56
#define DESCR_DVBSUBTITLES 0x59


#define PACK  __attribute__ ((__packed__))


// ETSI 300468 6.2.40
struct PACK ttxt_descr {
  uint8_t tag;
  uint8_t length;
  struct PACK {
    uint8_t lang[3];
    uint8_t type_mag; // 5 bits type, 3 bits mag
    uint8_t page_no;
  } d[1];
};

struct PACK sect_header {
  uint8_t table_id;
  uint16_t syntax_len;
  uint16_t transport_stream_id;
  uint8_t ver_cur;
  uint8_t section_number;
  uint8_t last_section_number;
};

// H.222.0 2.4.4.3, Table 2-25
struct PACK PAT_sect {
  uint8_t table_id;
  uint16_t syntax_len;
  uint16_t transport_stream_id;
  uint8_t vers_curr;
  uint8_t sect_no;
  uint8_t last_sect;
  struct PACK {
    uint16_t program_number;
    uint16_t res_PMTPID;
  } d[1];
};

// H.222.0 2.4.4.8, Table 2-28
struct PACK PMT_stream {
  uint8_t stream_type;
  uint16_t res_PID;
  uint16_t res_ES_info_len;
  uint8_t descrs[2];
};

struct PACK PMT_sect {
  uint8_t table_id;
  uint16_t syntax_len;
  uint16_t program_number;
  uint8_t vers_curr;
  uint8_t sect_no;
  uint8_t last_sect;
  uint16_t res_pcr;
  uint16_t res_program_info_length;
  struct PMT_stream s;
};


static int SetSectFilt(int fd, uint16_t pid, uint8_t tnr, uint8_t mask)
{
  int ret;

  struct dmx_sct_filter_params p;
  
  memset(&p, 0, sizeof(p));
  
  p.filter.filter[0] = tnr;
  p.filter.mask[0] = mask;
  p.pid     = pid;
  p.flags = DMX_IMMEDIATE_START | DMX_CHECK_CRC;
  
  if ((ret = ioctl(fd, DMX_SET_FILTER, &p)) < 0)  
    perror("DMX SET FILTER:");

  return ret;
}


static int
read_timeout(int fd, void *buf, size_t count, int timeout_ms) {
  int ret = -1;
  struct pollfd pi;

  pi.fd = fd;
  pi.events = POLLIN | POLLERR | POLLHUP | POLLNVAL;

  ret = poll(&pi, 1, timeout_ms);
  if(ret < 0)
    return errno;
  if(ret == 0) { // timeout
    esyslog("ttxtsubs: Service Information read: timeout!");
    return -1;
  }

  if(pi.revents == POLLIN) {
    ret = read(fd, buf, count);
    return ret;
  } else
    return -1;
}

/*
 * PID - pid to collect on
 * table_id - table id to filter out (H.222.0 table 2-26)
 * sects - pointer to char *[256], points to array of table sections, initially empty
 * numsects - number of sections in table
 * returns - 0 if everything ok
 */
#define SECTSIZE 1024
static int CollectSections(int card_no, int pid, int table_id, char **sects, int *numsects)
{
  int fd;
  int ret = -1;
  int last_section = 0;
  int done = 0;
  char *p = NULL;
  int n;
  char name[100];
  time_t start_time;

  // printf("CollectSections %d/%d\n", pid, table_id); // XXXX

  snprintf(name, sizeof(name), "/dev/dvb/adapter%d/demux0", card_no);

  memset(sects, 0, sizeof(char*) * 256);
  
  if((fd = open(name, O_RDWR)) < 0){
    perror("DEMUX DEVICE 1: ");
    return -1;
  }
  //  printf("ttxtsubs: siinfo.c: checking pid %d\n", pid); // XXXX
  if(SetSectFilt(fd, pid, table_id, 0xff)) {
    ret = -1;
    goto bail;
  }

  start_time = time(NULL);

  do {
    struct sect_header *h;
    int i;

    if((start_time + 1) < time(NULL)) {
      ret = -1;
      done = 1;
      break;
    }

    if(p == NULL)
      p = (char *) malloc(SECTSIZE);

    n = read_timeout(fd, p, SECTSIZE, 250);

    if(n < 8)
      continue;

    h = (struct sect_header *) p;

    if(n != ((ntohs(h->syntax_len) & 0xfff) + 3)) {
      esyslog("ttxtsubs: bad section length: %x / %x!\n", n, ntohs(h->syntax_len) & 0xfff);
      continue;
    }

    if(!(h->ver_cur & 0x01)) // if not current
      continue;

    last_section = h->last_section_number;

    if(!sects[h->section_number]) { // section_number
      sects[h->section_number] = p;
      p = NULL;
    }

    for(i = 0; i <= last_section; i++) {
      if(!sects[i]) {
	break;
      }

      if(i == last_section) {
	*numsects = last_section + 1;
	ret = 0;
	done = 1;
      }
    }

  } while (!done);

 bail:
  close(fd);
  return ret;
}

// When using a full featured card with hw_sections=0, at frequency change
// there seems to sometimes be left one (or more?) section block from the
// previous channel somewhere in the pipe. Whe need to remove that.
static void DiscardBufferedSections(int card_no, uint16_t pid, int table_id)
{
  int fd;
  char name[100];
  int ret = -1;
  int n = 0;
  char buf[SECTSIZE];

  snprintf(name, sizeof(name), "/dev/dvb/adapter%d/demux0", card_no);

  if((fd = open(name, O_RDWR)) < 0){
    perror("DEMUX DEVICE 1: ");
    return;
  }

  if(SetSectFilt(fd, pid, table_id, 0xff)) {
    goto bail;
  }

  // first read one section
  read_timeout(fd, buf, SECTSIZE, 1000);

  // this loop doesn't seem to be needed
  do {
    struct pollfd pi;
    pi.fd = fd;
    pi.events = POLLIN | POLLERR | POLLHUP | POLLNVAL;

    ret = poll(&pi, 1, 0);
    if(ret > 0 && (pi.revents & POLLIN)) {
      // TODO: Do something more useful if read fails!
      if(read(fd, buf, SECTSIZE) < 0) { }
      n++;
    }
  } while (ret > 0);

  if(n)
    printf("\nttxtsubs: DiscardBufferedSections: Discarded %d extra buffered sections\n\n", n + 1); // XXX
 bail:
  close(fd);
}

static void FreeSects(char **sects)
{
  int i;

  for(i = 0; i < 256; i++) {
    if(sects[i]) {
      free(sects[i]);
      sects[i] = NULL;
    }
  }
}


static void addpageinfo(struct ttxtinfo *info, uint16_t pid, struct ttxt_descr *descr, int descr_index)
{
  struct ttxtpidinfo *pd;
  struct ttxtpageinfo *pa;

  if(!info->p) {
    info->p = (struct ttxtpidinfo *) malloc(sizeof(*(info->p)));
    memset((char *) info->p, 0, sizeof(*(info->p)));
    info->p[0].pid = pid;
    info->pidcount++;
  }

  if(pid != info->p[info->pidcount - 1].pid) {
    info->pidcount++;
    info->p = (struct ttxtpidinfo *) realloc(info->p, sizeof(info->p[0]) * info->pidcount);
    memset((char *) &(info->p[info->pidcount - 1]), 0, sizeof(info->p[0]));
    info->p[info->pidcount - 1].pid = pid;
  }

  pd = &(info->p[info->pidcount - 1]);

  pd->pagecount++;
  if(!pd->i) {
    pd->i = (struct ttxtpageinfo *) malloc(sizeof(pd->i[0]));
  } else {
    pd->i = (struct ttxtpageinfo *) realloc(pd->i, sizeof(pd->i[0]) * pd->pagecount);
  }
  memset((char *) &(pd->i[pd->pagecount - 1]), 0, sizeof(pd->i[0]));

  pa = &(pd->i[pd->pagecount - 1]);

  pa->lang[0] = descr->d[descr_index].lang[0];
  pa->lang[1] = descr->d[descr_index].lang[1];
  pa->lang[2] = descr->d[descr_index].lang[2];
  pa->type = descr->d[descr_index].type_mag >> 3;
  pa->mag = descr->d[descr_index].type_mag & 0x7;
  pa->page = descr->d[descr_index].page_no;

  if(globals.frenchSpecial()) {
    /* By some silly reason some French channels (Canal+ and CanalSatellite)
       announce subtitles with the page number in decimal, for example
       page 859 when they really are on page 889. */
    uint8_t newpage = pa->page;
    if(pa->page >= 0x50 && pa->page <= 0x59)
      newpage += 0x30;
    if(pa->page != newpage) {
      uint8_t mag = pa->mag;
      if(mag == 0)
	mag = 8;
      esyslog("ttxtsubs: Warning: Remapped page number %01x%02x to %01x%02x!\n",
	      mag, pa->page, mag, newpage);
      pa->page = newpage;
    }
  }
}


static int HasVPID(int vpid, char **pmtsects, int numsects)
{
  int found = 0;
  int i;

  for(i = 0; i < numsects && !found; i++) {
    struct PMT_sect *psect;
    char *sp;
    char *end;

    psect = (struct PMT_sect *) pmtsects[i];
    end = pmtsects[i] + (ntohs(psect->syntax_len) & 0x3ff) - 7;
    // skip extra program info
    sp = ((char *) &(psect->s)) + (ntohs(psect->res_program_info_length) & 0xfff);
    
    while(sp < end && !found) {
      struct PMT_stream *s = (struct PMT_stream *) sp;

      if(s->stream_type == 1 || s->stream_type == 2) { // Video stream
	if((ntohs(s->res_PID) & 0x1fff) == vpid) {
	  //printf("Found ttxt by vpid: %d/%d\n", s->stream_type, (ntohs(s->res_PID) & 0x1fff));
	  found = 1;
	  break;
	}
      }
      sp += (ntohs(s->res_ES_info_len) & 0xfff) + 5;
    }
  }  
  
  return found;
}

static void ExtractTtxtInfo(char **pmtsects, int numsects, struct ttxtinfo *info, int *foundinfo)
{
  int i;

  for(i = 0; i < numsects; i++) {
    struct PMT_sect *psect;
    char *sp;
    char *end;

    psect = (struct PMT_sect *) pmtsects[i];
    end = pmtsects[i] + (ntohs(psect->syntax_len) & 0x3ff) - 7;
    // skip extra program info
    sp = ((char *) &(psect->s)) + (ntohs(psect->res_program_info_length) & 0xfff);
    
    while(sp < end) {
      struct PMT_stream *s = (struct PMT_stream *) sp;

      if(s->stream_type == 6) { // PES private data
	uint8_t *descr;

	for(descr = s->descrs; descr < s->descrs + (ntohs(s->res_ES_info_len) & 0xfff); descr += descr[1] + 2) {
	  if(descr[0] == DESCR_TELETEXT) {
	    struct ttxt_descr *t = (struct ttxt_descr *) descr;
	    int i, count = t->length / sizeof(t->d[0]);
	    for(i = 0; i < count; i++) {
	      addpageinfo(info, ntohs(s->res_PID) & 0x1fff, t, i);
	      //printf("%c%c%c: type %d, page: %01x%02x\n", t->d[i].lang[0], t->d[i].lang[1], t->d[i].lang[2], t->d[i].type_mag >> 3, t->d[i].type_mag & 0x7,  t->d[i].page_no);
	      *foundinfo = 1;
	    }
	  }
	}
      }
      sp += (ntohs(s->res_ES_info_len) & 0xfff) + 5;
    }
  }  
}

/*
 * if vpid != 0, we first search if we can find the vpid in this PMT,
 * and if so we go get the ttxt pid
 * return <> 0 on error
 */
static int FindTtxtInfoInPMT(int card_no, int pid, int vpid, struct ttxtinfo *info, int *foundinfo)
{
  int ret = -1;
  char *pmtsects[256];
  int numsects;

  // printf("FindTtxtInfoInPMT pid: %d, vpid: %d\n", pid, vpid); // XXXX

  ret = CollectSections(card_no, pid, 0x02, pmtsects, &numsects);
  if(ret)
    goto bail;

  if(vpid != 0) {
    if(!HasVPID(vpid, pmtsects, numsects)) {
      goto bail;
    }
  }

  ExtractTtxtInfo(pmtsects, numsects, info, foundinfo);

 bail:
  FreeSects(pmtsects);
  return ret;
}


/*
 * find the ttxt_info in the PMT via the PAT, try first with the SID
 * and if that fails with the VPID
 * return <> 0 on error;
 */
int GetTtxtInfo(int card_no, const cChannel *c, struct ttxtinfo *info)
{
  int ret = -1;
  char *patsects[256];
  int numsects;
  int i;
  int j;
  uint16_t pmt_pid = 0;
  int foundinfo = 0;
  int retry;

  memset((char *) info, 0, sizeof(*info));

#if 1
  // <manual page selection patch>
  cTtxtSubsChannelSetting *cs = 
    TtxtSubsChannelSettings.Get(c);
    
  if (cs && cs->PageMode() == PAGE_MODE_MANUAL) {
    struct ttxt_descr descr;
 
    if (cs->PageNumber() > 0) {
      int temp = cs->PageNumber();
      if (temp>=800) temp-=800;
      char langs[MAXLANGUAGES][2][4];
      char hi[MAXLANGUAGES][2];
      //put as the first language the first one configured,
      //so that FindSubs will later accept it
      memcpy(langs,globals.languages(),sizeof(langs));
      memcpy(hi,globals.hearingImpaireds(),sizeof(hi));
      for (i = 0; i < MAXLANGUAGES; i++) {
        for (j = 0; j < 2 ; j++) {
          if (!langs[i][j][0]) continue;
         
          descr.d[0].lang[0] = langs[i][j][0];
          descr.d[0].lang[1] = langs[i][j][1];
          descr.d[0].lang[2] = langs[i][j][2];
          if (hi[i][j]) 
            descr.d[0].type_mag = (TTXT_SUBTITLE_HEARING_IMPAIRED_PAGE << 3) + (temp / 100); 
          else 
            descr.d[0].type_mag = (TTXT_SUBTITLE_PAGE << 3) + (temp / 100);

          temp = temp % 100;
          descr.d[0].page_no = ((temp / 10) << 4) + (temp % 10);
      
          addpageinfo(info, c->Tpid(), &descr, 0);
          return 0;
        }
      }
      //no language configured
      return -1;       
    }
  }
  if (cs && cs->PageMode() == PAGE_MODE_DISABLED) {
    return -1;
  }
  // </manual page selection patch>
#endif

  for(retry = 0; retry <= 1 && !foundinfo; retry++) { // XXX retry two times due to flaky pat scanning with hw_sections=0

    // printf("GetTtxtInfo A sid: %d, vpid: %d\n", sid, vpid); // XXXX

#if 1
    // firmware fixed now?
    DiscardBufferedSections(card_no, 0, 0);
#endif
    
    ret = CollectSections(card_no, 0, 0, patsects, &numsects);
    if(ret)
      goto bail;
    
    if(c->Sid() != 0) {
      int found;
      
      for(i = 0, found = 0; i < numsects && !found; i++) {
	int numdescrs;
	struct PAT_sect *s = (struct PAT_sect *) patsects[i];
	
	numdescrs = ((ntohs(s->syntax_len) & 0x3FF) - 7) / 4;
	
	for(j = 0; j < numdescrs && !found; j++) {
	  uint16_t pno = ntohs(s->d[j].program_number);
	  
	  if(pno == 0)
	    continue; // network pid
	  
	  if(pno == c->Sid()) {
	    pmt_pid = ntohs(s->d[j].res_PMTPID) & 0x1fff;
	    found = 1;
	  }
	}
      }
    }
    
    if(pmt_pid != 0) {
      // printf("GetTtxtInfo B pmt_pid: %d, vpid: %d\n", pmt_pid, vpid); // XXXX
      ret = FindTtxtInfoInPMT(card_no, pmt_pid, 0, info, &foundinfo);
    } else {
      // SID not found, try searching VID in all SIDS
      if(c->Vpid() != 0) {
	int done;
	for(i = 0, done = 0; i < numsects && !done; i++) {
	  int numdescrs;
	  struct PAT_sect *s = (struct PAT_sect *) patsects[i];
	  
	  numdescrs = ((ntohs(s->syntax_len) & 0x3FF) - 7) / 4;
	  
	  for(j = 0; j < numdescrs && !done; j++) {
	    uint16_t pno = ntohs(s->d[j].program_number);
	    
	    if(pno == 0)
	      continue; // network pid
	    
	    pmt_pid = ntohs(s->d[j].res_PMTPID) & 0x1fff;
	    
	    // printf("GetTtxtInfo C pmt_pid: %d, vpid: %d\n", pmt_pid, vpid); // XXXX
	    ret = FindTtxtInfoInPMT(card_no, pmt_pid, c->Vpid(), info, &foundinfo);
	    if(ret) {
	      done = 1;
	    }
	    if(foundinfo)
	      done = 1;
	  }
	}
      }
    }

    FreeSects(patsects);
  }

bail:
  return ret;
}


void FreeTtxtInfoData(struct ttxtinfo *info)
{
  int i;

  for(i = 0; i < info->pidcount; i++) {
    free((void *) info->p[i].i);
  }

  free((void *) info->p);
}


void DupTtxtInfo(struct ttxtinfo *in, struct ttxtinfo *out)
{
  int i;

  out->pidcount = in->pidcount;
  out->p = (struct ttxtpidinfo *) malloc(sizeof(out->p[0]) * in->pidcount);

  for(i = 0; i < in->pidcount; i++) {
    out->p[i] = in->p[i];
    out->p[i].i = (struct ttxtpageinfo *) malloc(sizeof(out->p[0].i[0]) * in->p[i].pagecount);
    memcpy((void *) out->p[i].i, (void *) in->p[i].i, sizeof(out->p[0].i[0]) * in->p[i].pagecount);
  }
}


struct ttxtpidinfo *FindSubs(struct ttxtinfo *info, int *pid, int *pageno, char *lang)
{
  struct ttxtpidinfo *foundPI = NULL;
  int foundChoise = 1000;

  // walk through all available languages and remember most preferred one
  for(int i = 0; i < info->pidcount; i++) {
    for(int j = 0; j < info->p[i].pagecount; j++) {
      if((info->p[i].i[j].type != TTXT_SUBTITLE_PAGE) &&
	 (info->p[i].i[j].type != TTXT_SUBTITLE_HEARING_IMPAIRED_PAGE))
	continue;

      int ch = globals.langChoise(info->p[i].i[j].lang,
	  info->p[i].i[j].type == TTXT_SUBTITLE_HEARING_IMPAIRED_PAGE);
      if(ch != -1) {
	if(ch < foundChoise) {
	  foundChoise = ch;
	  *pid = info->p[i].pid;
	  *pageno = info->p[i].i[j].mag * 0x100 + info->p[i].i[j].page;
	  strncpy(lang, info->p[i].i[j].lang, 3);
	  foundPI = &(info->p[i]);
	}
      }
    }
  }

  if(foundPI) {
    return foundPI;
  }

  if(info->pidcount == 0)
    isyslog("ttxtsubs: No teletext subtitles on channel.");
  else {
    isyslog("ttxtsubs: Wanted subtitle language(s) not found on channel, available languages:");
    for(int i = 0; i < info->pidcount; i++) {
      for(int j = 0; j < info->p[i].pagecount; j++) {
	int page = info->p[i].i[j].mag * 0x100 + info->p[i].i[j].page;
	int type = info->p[i].i[j].type;
	if(page < 0x100)
	  page += 0x800;
	isyslog("ttxtsubs:     %03x: %c%c%c %s", page, info->p[i].i[j].lang[0],
		info->p[i].i[j].lang[1], info->p[i].i[j].lang[2],
		type == TTXT_INITIAL_PAGE ? "(Initial Page (The teletext start page, not a subtitles page!))" :
		type == TTXT_SUBTITLE_PAGE ? "(Subtitles)" :
		type == TTXT_ADDITIONAL_INFO_PAGE ? "(Additional Info Page)" :
		type == TTXT_PROGRAMME_SCHEDULE_PAGE ? "(Programme Schedule Page)" :
		type == TTXT_SUBTITLE_HEARING_IMPAIRED_PAGE ? "(Hearing Impaired)" : "(Unknown type)");
      }
    }
  }

  *pid = 0;
  *pageno = -1;
  return NULL;
}
